pex/resolve/downloads.py

```diff
@@ -45,10 +48,20 @@ def get_downloads_dir(pex_root=None):
 @attr.s(frozen=True)
 class ArtifactDownloader(object):
     resolver = attr.ib()  # type: Resolver
-    pip = attr.ib(factory=get_pip)  # type: Pip
+    target = attr.ib(default=LocalInterpreter.create())  # type: Target
     package_index_configuration = attr.ib(
         default=PackageIndexConfiguration.create()
     )  # type: PackageIndexConfiguration
+    max_parallel_jobs = attr.ib(default=None)  # type: Optional[int]
+    pip = attr.ib(init=False)  # type: Pip
+
+    @pip.default
```

Comment:
Took me a little spelunking of the attrs docs to grok this, and particularly the use of the partially-initialized self, but it's neat!

Reply:
Yes, attrs is quite a bit nicer than dataclasses.